src/sxg_cert_chain.c

```diff
@@ -0,0 +1,343 @@
+  } else {
+    rv = 1;
+  }
+  return rv != 0 &&  // Timeout on request.
+         rv != -1;  // Select error.
+}
+
+bool sxg_execute_ocsp_request(BIO* io, const char* path,
+                              OCSP_CERTID* id, OCSP_RESPONSE** dst) {
+  int fd;
+  if (BIO_get_fd(io, &fd) < 0) {  // Can't get connection fd.
+    return false;
+  }
+  OCSP_REQ_CTX* const octx = OCSP_sendreq_new(io, path, NULL, -1);
+  OCSP_REQUEST* const req = OCSP_REQUEST_new();
+  bool success = OCSP_request_add0_id(req, id) &&
+                 OCSP_REQ_CTX_set1_req(octx, req) &&
+                 wait_fd(fd, false, true);
+  while (success) {
+    if (OCSP_sendreq_nbio(dst, octx) != -1) {
+      break;
+    }
+    success = success && wait_fd(fd, BIO_should_read(io), BIO_should_write(io));
+  }
+  
+  OCSP_REQUEST_free(req);
+  OCSP_REQ_CTX_free(octx);
+  return success;
+}
+
+static bool sxg_make_ocsp_session(const char* ocsp_url,
+                                  char** path,
+                                  BIO** connection) {
+  char* host;
+  char* port;
+  int use_ssl;
+  OCSP_parse_url(ocsp_url, &host, &port, path, &use_ssl);
+  *connection = BIO_new_connect(host);
+
+  bool success = *connection != NULL;
+  if (success && port != NULL) {
+    BIO_set_conn_port(*connection, port);
+  }
+
+  if (use_ssl == 1) {
+    SSL_CTX* const ssl_ctx = SSL_CTX_new(SSLv23_client_method());
+    if (ssl_ctx == NULL) {
+      success = false;;
+    } else {
+      SSL_CTX_set_mode(ssl_ctx, SSL_MODE_AUTO_RETRY);
+      *connection = BIO_push(BIO_new_ssl(ssl_ctx, 1), *connection);
+    }
+  }
+
+  OPENSSL_free(host);
+  OPENSSL_free(port);
+  return success;
+}
+
+bool sxg_fetch_ocsp_response(X509* cert, X509* issuer, OCSP_RESPONSE** dst) {
+  sxg_buffer_t ocsp_url = sxg_empty_buffer();
+  char* path = NULL;
+  BIO* cbio = NULL;
+  bool success =
+      sxg_extract_ocsp_url(cert, &ocsp_url) &&
+      sxg_make_ocsp_session((const char*)ocsp_url.data, &path, &cbio) &&
+      BIO_do_connect(cbio) > 0 &&
+      sxg_execute_ocsp_request(cbio,
+                               path,
+                               OCSP_cert_to_id(EVP_sha256(), cert, issuer),
+                               dst);
+
+  sxg_buffer_release(&ocsp_url);  
+  OPENSSL_free(path);
+  BIO_free(cbio);
+  return success;
+}
+
+bool sxg_cert_chain_append_cert(X509* cert,
+                                OCSP_RESPONSE* ocsp_response,
+                                const sxg_buffer_t* sct_list,
+                                sxg_cert_chain_t* target) {
+  X509_up_ref(cert);
+  if (!ensure_free_capacity(1, target)) {
+    return false;
+  }
+  sxg_cert_t* const new_cert = &target->certs[target->size];
+  new_cert->certificate = cert;
+  new_cert->ocsp_response = ocsp_response;
+  new_cert->sct_list = sxg_empty_buffer();
+  sxg_buffer_copy(sct_list, &new_cert->sct_list);
+  target->size++;
+  return true;
+}
+
+static bool sxg_write_array_cbor_header(uint64_t length, sxg_buffer_t* target) {
```

Comment:
Shouldn't this function be in sxg_buffer.c?

Reply:
Optional: also *_cbor_header functions can be unified as follows:
```
// Major type number goes in the first 3 bits.
const uint8_t kByteStringMagic = 2 << 5;
const uint8_t kTextStringMagic = 3 << 5;
const uint8_t kArrayMagic = 4 << 5;
...

const uint8_t kUint8Magic = 24;
const uint8_t kUint16Magic = 25;
...

bool sxg_write_cbor_header(uint8_t major_type_magic, uint64_t length, sxg_buffer_t* target) {
  if (length <= 0x17) {
    return sxg_write_byte(major_type_magic + length, target)
  } else if (length <= 0xff) {
    return sxg_write_byte(major_type_magic + kUint8Magic) && ...
  } ...
  ...
}
```

Reply:
In this library, cbor_header writer is defined where it is used., like https://github.com/google/libsxg/blob/master/src/sxg_header.c#L173-L187
I thought it is not confusing because the use case is limited. But this time I am using them in broader way.
So extract and merge them into some header file like `sxg_cbor.h` will be the best choice, but it should be done in another patch.

Reply:
https://github.com/google/libsxg/issues/35
issue created.